### PR TITLE
Create common code for the variety of Issue queries

### PR DESCRIPTION
**Pull request by BillWagner**
As a precursor to being able to sequester PRs, create some common code for processing queries for issues.

Contributes to #266 

The changes fall under three different umbrellas:

1. **Update to .NET 8**: Update all projects to .NET 8, update NuGet packages to the latest .NET 8 packages.
2. **Write tests for Quest issue readers**: Since a lot of the quest issue code was reworked in this PR, write tests for processing nodes with a variety of different conditions.
3. **Create common library for querying issues**: This factors the common code into a single set of utilities, and creates generic interfaces that allow for specialization on each type.

The common library is a work in progress. I expect further updates in the next precursor PR for creating common code to handle all queries for PRs.

### CODE REVIEW

File: DotNet.DocsTools/GitHubObjects/QuestIssue.cs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+        this.ClosingPRUrl = ResponseExtractors.GetChildArrayElements(issueNode, "timelineItems", item =>
+            (item.TryGetProperty("closer", out var closer) && closer.ValueKind == JsonValueKind.Object) ?
+            ResponseExtractors.StringProperty(closer, "url")
+            : default).LastOrDefault(url => url is not null);
+
+        this.LinkText = $"""
+        <a href = "https://github.com/{organization}/{repository}/issues/{Number}">
+            {organization}/{repository}#{Number}
+        </a>
+        """;
```
Comment (IEvangelist) [inline DotNet.DocsTools/GitHubObjects/QuestIssue.cs:288]:
```suggestion
        FormattedAuthorLoginName = (author is not null) ?
            $"{author.Login} - {author.Name}" :  
            "Ghost - ";

        IsOpen = ResponseExtractors.StringProperty(issueNode, "state") is "OPEN";
        BodyHtml = ResponseExtractors.StringProperty(issueNode, "bodyHTML");
        UpdatedAt = ResponseExtractors.GetUpdatedAtValueOrNow(issueNode);

        Assignees = [ ..ResponseExtractors.GetChildArrayElements(issueNode, "assignees", item =>
            Actor.FromJsonElement(item)).Where(actor => actor is Actor)];
 
        Labels = ResponseExtractors.GetChildArrayElements(issueNode, "labels", item => new GitHubLabel(item)); 
        Comments = ResponseExtractors.GetChildArrayElements(issueNode, "comments", item =>
        {
            var actor = Actor.FromJsonElement(ResponseExtractors.GetAuthorChildElement(item));
            return ((actor is not null) ? actor.Login : "Ghost", 
                ResponseExtractors.StringProperty(item, "bodyHTML"));
            }
        ).ToArray();

        var points = ResponseExtractors.GetChildArrayElements(issueNode, "projectItems", item =>
            StoryPointSize.OptionalFromJsonElement(item));
        ProjectStoryPoints = [ ..points.Where(sz => sz is not null).ToArray()];

        // check state. If re-opened, don't reference the (not correct) closing PR
        ClosingPRUrl = ResponseExtractors.GetChildArrayElements(issueNode, "timelineItems", item =>
            (item.TryGetProperty("closer", out var closer) && closer.ValueKind == JsonValueKind.Object) ?
            ResponseExtractors.StringProperty(closer, "url")
            : default).LastOrDefault(url => url is not null);

        LinkText = $"""
        <a href = "https://github.com/{organization}/{repository}/issues/{Number}">
            {organization}/{repository}#{Number}
        </a>
        """;
```

File: DotNet.DocsTools/GitHubObjects/QuestIssue.cs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    /// </summary>
+    public string? ClosingPRUrl { get; }
+
+    /// <summary>
+    /// Retrieve the assigned name, if an MS employee
+    /// </summary>
+    /// <param name="ospoClient">The Open Source program office client service.</param>
+    /// <returns>The email address of the assignee. Null if unassigned, or the assignee is not a 
+    /// Microsoft FTE.</returns>
+    public async Task<string?> AssignedMicrosoftEmailAddress(OspoClient ospoClient)
```
Comment (IEvangelist) [inline DotNet.DocsTools/GitHubObjects/QuestIssue.cs:350]: I'd recommend changing the name of this method to indicate that it's doing network I/O. Perhaps add a verb, like `Get` and a suffix like `Async`.

File: DotNet.DocsTools/GitHubObjects/StoryPointSize.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+using System.Text.Json;
+
+namespace DotNet.DocsTools.GitHubObjects;
+
+/// <summary>
+/// Record type to store a project name / story point size pair
+/// </summary>
+public record StoryPointSize
+{
+
```
Comment (IEvangelist) [inline DotNet.DocsTools/GitHubObjects/StoryPointSize.cs:11]:
```suggestion
```

File: DotNet.DocsTools/GitHubObjects/StoryPointSize.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this.CalendarYear = CalendarYear;
+        this.Month = Month;
+        this.Size = Size;
+    }
+
+    public int CalendarYear { get; }
+    public string Month { get; }
+    public string Size { get; }
+}
+
```
Comment (IEvangelist) [inline DotNet.DocsTools/GitHubObjects/StoryPointSize.cs:58]:
```suggestion
```

File: DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/// <typeparam name="TResult">The type of the result objects.</typeparam>
+/// <typeparam name="TVariables">A record type containing the variables of the query. </typeparam>
+/// <remarks>
+/// This generic type performs a query of GitHub's GraphQL endpoint, and returns an async
+/// enumerable of the result objects. The query is parameterized by a record type, and the
+/// result type. The result type must implement the <see cref="IGitHubQueryResult{TResult, TVariables}"/>
+/// </remarks>
+public class EnumerationQuery<TResult, TVariables> where TResult : IGitHubQueryResult<TResult, TVariables>
+{
+    private readonly IGitHubClient client;
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs:18]:
```suggestion
    private readonly IGitHubClient _client;
```

File: DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    /// <summary>
+    /// Construct the query object.
+    /// </summary>
+    /// <param name="client">The GitHub client.</param>
+    public EnumerationQuery(IGitHubClient client)
+    {
+        this.client = client ?? throw new ArgumentNullException(paramName: nameof(client), message: "Cannot be null");
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs:26]:
```suggestion
        _client = client ?? throw new ArgumentNullException(paramName: nameof(client), message: "Cannot be null");
```

File: DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+/// <typeparam name="TResult">The type of the result objects.</typeparam>
+/// <typeparam name="TVariables">A record type containing the variables of the query. </typeparam>
+/// <remarks>
+/// This generic type performs a query of GitHub's GraphQL endpoint, and returns an async
+/// enumerable of the result objects. The query is parameterized by a record type, and the
+/// result type. The result type must implement the <see cref="IGitHubQueryResult{TResult, TVariables}"/>
+/// </remarks>
+public class ScalarQuery<TResult, TVariables> where TResult : IGitHubQueryResult<TResult, TVariables>
+{
+    private readonly IGitHubClient client;
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs:19]:
```suggestion
    private readonly IGitHubClient _client;
```

File: DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    /// <summary>
+    /// Construct the query object.
+    /// </summary>
+    /// <param name="client">The GitHub client.</param>
+    public ScalarQuery(IGitHubClient client)
+    {
+        this.client = client ?? throw new ArgumentNullException(paramName: nameof(client), message: "Cannot be null");
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs:27]:
```suggestion
        _client = client ?? throw new ArgumentNullException(paramName: nameof(client), message: "Cannot be null");
```

File: DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    public async IAsyncEnumerable<TResult> PerformQuery(TVariables variables)
+    {
+        var findIssuesPacket = TResult.GetQueryPacket(variables);
+
+        var cursor = default(string);
+        bool hasMore = true;
+        while (hasMore)
+        {
+            findIssuesPacket.variables["cursor"] = cursor!;
+            var jsonData = await client.PostGraphQLRequestAsync(findIssuesPacket);
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/EnumerationQuery.cs:46]:
```suggestion
            var jsonData = await _client.PostGraphQLRequestAsync(findIssuesPacket);
```

File: DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    /// <returns>The async enumerable.</returns>
+    /// <remarks>
+    /// This query encapsulates the paging API for GitHub's GraphQL 
+    /// endpoint.
+    /// </remarks>
+    public async Task<TResult> PerformQuery(TVariables variables)
+    {
+        var scalarPacket = TResult.GetQueryPacket(variables);
+
+        var rootElement= await client.PostGraphQLRequestAsync(scalarPacket);
```
Comment (IEvangelist) [inline DotNet.DocsTools/GraphQLQueries/ScalarQuery.cs:42]:
```suggestion
        var rootElement= await _client.PostGraphQLRequestAsync(scalarPacket);
```

**Review by IEvangelist — APPROVED** [review]
LGTM, that's a lot of changes... We'll need to watch this closely once merged.

File: actions/docs-verifier/src/ActionRunner/ActionRunner.csproj [hunk 11]
```diff
@@ -1,8 +1,10 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net7.0</TargetFramework>
+    <TargetFramework>net8</TargetFramework>
+    <nulable>enable</nulable>
```
Comment (gewarren) [inline actions/docs-verifier/src/ActionRunner/ActionRunner.csproj:6]:
```suggestion
    <nullable>enable</nullable>
```

**Review by gewarren — APPROVED** [review]
Just skimmed it but looks good.

**Pull request merged**